src/go/pgo/distsys/distsys.go

```diff
@@ -1,7 +1,388 @@
+}
+
+func (e KeyNotFoundError) Error() string {
+	return fmt.Sprintf("Key not found error [%v]", string(e))
+}
+
+func (e RangeMismatchError) Error() string {
+	return fmt.Sprintf("Range mismatch error")
+}
+
+func (e TransactionError) Error() string {
+	return fmt.Sprintf("Transaction error [%v]", e)
+}
+
+type AcquireSet struct {
+	ReadNames  []string
+	WriteNames []string
+}
+
+type ReleaseSet struct {
+	WriteNames  []string
+	Transaction TransactionSet
+}
+
+type StateServer interface {
+	Acquire(set *AcquireSet) (*ReleaseSet, map[string]interface{}, error)
+	Release(set *ReleaseSet, values ...interface{}) error
+	WaitPeers()
+
+	SetPolicy(policy IMigrationPolicy)
+
+	Close() error
+}
+
+type IMigrationPolicy interface {
+	OnGet(host string, key string)
+	OnSet(host string, key string)
+	MigrateTo(host string, key string) bool
+}
+
+type LocalDataStore interface {
+	Get(key string) (interface{}, error)
+	GetExclusive(key string) (interface{}, error)
+	Set(key string, value interface{})
+	Release(key string)
+
+	remove(key string)
+	create(key string, value interface{})
+}
+
+func (t *MostFrequentlyUsed) OnGet(host string, key string) {
+	entry, ok := t.Keys[key]
+
+	if !ok {
+		entry = HostFrequency{
+			Frequency: make(map[string]int),
+		}
+	}
+
+	freq, ok := entry.Frequency[host]
+	if !ok {
+		freq = 0
+	}
+
+	freq++
+
+	entry.Frequency[host] = freq
+	t.Keys[key] = entry
+}
+
+func (t *MostFrequentlyUsed) OnSet(host string, key string) {
+	t.OnGet(host, key)
+}
+
+func (t *MostFrequentlyUsed) MigrateTo(host string, key string) bool {
+	entry, ok := t.Keys[key]
+
+	if !ok {
+		return false
+	}
+
+	var maxHost string
+	var maxFreq int
+
+	for curr, freq := range entry.Frequency {
+		if freq > maxFreq {
+			maxHost = curr
+			maxFreq = freq
+		}
+	}
+
+	if host != maxHost {
+		return false
+	}
+
+	return true
+}
+
+func NewStateServer(peers []string, self, coordinator string, initValues map[string]interface{}) (StateServer, error) {
+	log.SetFlags(log.Ldate | log.Ltime | log.Lshortfile)
+
+	pi := NewProcessInitialization(peers, self, coordinator)
+
+	// FIXME this is assuming everything is centralized in one place
+	selfId := -1
+	coordinatorId := -1
+	for i, p := range peers {
+		if p == self {
+			selfId = i
+		}
+
+		if p == coordinator {
+			coordinatorId = i
+		}
+	}
+	// Make sure `self` is in the list of peers
+	if selfId < 0 {
+		panic("self is not in peers")
+	}
+	if coordinatorId < 0 {
+		panic("coodinator is not in peers")
+	}
+
+	owners := make(map[string]*ObjectOwner, len(initValues))
+	store := make(map[string]interface{})
+
+	// at first, all state is in the coordinator node
+	for key, _ := range initValues {
+		owners[key] = &ObjectOwner{
+			hostID: coordinatorId,
+		}
+	}
+
+	if pi.isCoordinator() {
+		store = initValues
+	}
+
+	network := &Network{
+		ProcessInitialization: pi,
+		store:     CreateSimpleDataStore(store),
+		owners:    owners,
+		localhost: self,
+		hosts:     peers,
+		migration: &NeverMigrate{},
```

Comment:
I think we should switch this to AlwaysMigrate. Thoughts?

Reply:
This should actually be configurable. But yeah, I think it had better performance.